File: kr_stock_trader/history.py
"""체결 기록(JSON Lines)과 통계."""

import json
import os
from dataclasses import asdict, dataclass
from datetime import datetime
from typing import List, Optional

from .market import KST, now_kst


@dataclass
class TradeRecord:
    time: str
    symbol: str
    side: str
    qty: int
    price: float
    reason: str = ""
    pnl: Optional[float] = None   # 매도(청산) 시 실현손익 (수수료·세금 차감)
    order_id: str = ""
    broker: str = ""


class TradeHistoryStore:
    def __init__(self, path: str):
        self.path = path

    def append(self, record: TradeRecord) -> None:
        folder = os.path.dirname(os.path.abspath(self.path))
        os.makedirs(folder, exist_ok=True)
        with open(self.path, "a", encoding="utf-8") as f:
            f.write(json.dumps(asdict(record), ensure_ascii=False) + "\n")

    def load(self) -> List[TradeRecord]:
        if not os.path.exists(self.path):
            return []
        out = []
        with open(self.path, encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if line:
                    out.append(TradeRecord(**json.loads(line)))
        return out

    def closed_trades(self) -> List[TradeRecord]:
        return [r for r in self.load() if r.pnl is not None]

    def realized_pnl_on(self, day: datetime = None) -> float:
        key = (day or now_kst()).astimezone(KST).strftime("%Y-%m-%d")
        return sum(r.pnl for r in self.closed_trades() if r.time.startswith(key))

    def consecutive_losses(self) -> tuple:
        """(최근 연속 손실 횟수, 마지막 손실 시각)"""
        count, last_time = 0, None
        for r in reversed(self.closed_trades()):
            if r.pnl < 0:
                count += 1
                last_time = last_time or datetime.fromisoformat(r.time)
            else:
                break
        return count, last_time

    def stats(self) -> dict:
        closed = self.closed_trades()
        wins = [r for r in closed if r.pnl > 0]
        losses = [r for r in closed if r.pnl <= 0]
        gross_win = sum(r.pnl for r in wins)
        gross_loss = -sum(r.pnl for r in losses)
        return {
            "trades": len(closed),
            "wins": len(wins),
            "losses": len(losses),
            "win_rate": (len(wins) / len(closed) * 100.0) if closed else 0.0,
            "total_pnl": gross_win - gross_loss,
            "profit_factor": (gross_win / gross_loss) if gross_loss else None,
        }

File: kr_stock_app/lib/core/history.dart
// 체결 기록과 통계.
import 'dart:convert';

class TradeRecord {
  final DateTime time; // KST
  final String symbol, side, reason, orderId, broker;
  final int qty;
  final double price;
  final double? pnl;
  const TradeRecord({
    required this.time,
    required this.symbol,
    required this.side,
    required this.qty,
    required this.price,
    this.reason = '',
    this.pnl,
    this.orderId = '',
    this.broker = '',
  });

  Map<String, dynamic> toJson() => {
        't': time.toIso8601String(),
        's': symbol,
        'd': side,
        'q': qty,
        'p': price,
        'r': reason,
        'l': pnl,
        'o': orderId,
        'b': broker,
      };

  factory TradeRecord.fromJson(Map<String, dynamic> j) => TradeRecord(
        time: DateTime.parse(j['t'] as String),
        symbol: j['s'] as String,
        side: j['d'] as String,
        qty: (j['q'] as num).toInt(),
        price: (j['p'] as num).toDouble(),
        reason: j['r'] as String? ?? '',
        pnl: (j['l'] as num?)?.toDouble(),
        orderId: j['o'] as String? ?? '',
        broker: j['b'] as String? ?? '',
      );
}

class TradeStats {
  final int trades, wins, losses;
  final double winRate, totalPnl;
  final double? profitFactor;
  const TradeStats(this.trades, this.wins, this.losses, this.winRate, this.totalPnl, this.profitFactor);
}

class TradeHistory {
  final List<TradeRecord> records;
  final void Function(List<TradeRecord>)? onChanged;
  TradeHistory([List<TradeRecord>? initial, this.onChanged]) : records = initial ?? [];

  void add(TradeRecord r) {
    records.add(r);
    if (records.length > 2000) records.removeRange(0, records.length - 2000);
    onChanged?.call(records);
  }

  void clear() {
    records.clear();
    onChanged?.call(records);
  }

  Iterable<TradeRecord> get closed => records.where((r) => r.pnl != null);

  double realizedOn(DateTime kst) =>
      closed.where((r) => _sameDay(r.time, kst)).fold(0.0, (a, r) => a + r.pnl!);

  /// (연속 손실 횟수, 마지막 손실 시각)
  (int, DateTime?) consecutiveLosses() {
    var count = 0;
    DateTime? last;
    for (final r in closed.toList().reversed) {
      if (r.pnl! < 0) {
        count++;
        last ??= r.time;
      } else {
        break;
      }
    }
    return (count, last);
  }

  TradeStats stats() {
    final c = closed.toList();
    final wins = c.where((r) => r.pnl! > 0).toList();
    final losses = c.where((r) => r.pnl! <= 0).toList();
    final gw = wins.fold(0.0, (a, r) => a + r.pnl!);
    final gl = -losses.fold(0.0, (a, r) => a + r.pnl!);
    return TradeStats(c.length, wins.length, losses.length, c.isEmpty ? 0 : wins.length / c.length * 100, gw - gl,
        gl > 0 ? gw / gl : null);
  }

  String encode() => jsonEncode(records.map((e) => e.toJson()).toList());
  static List<TradeRecord> decode(String? s) {
    if (s == null || s.isEmpty) return [];
    try {
      return (jsonDecode(s) as List).map((e) => TradeRecord.fromJson((e as Map).cast<String, dynamic>())).toList();
    } catch (_) {
      return [];
    }
  }
}

bool _sameDay(DateTime a, DateTime b) => a.year == b.year && a.month == b.month && a.day == b.day;
